theseus/labs/embodied/robot/forward_kinematics.py

```diff
@@ -123,15 +109,72 @@ def backward(cls, ctx, *grad_outputs):
     return (
         ForwardKinematics,
         _forward_kinematics_impl,
-        _jforward_kinematics_impl,
         _forward_kinematics_helper,
-        _jforward_kinematics_helper,
+        _forward_kinematics_backward_helper,
+        link_ids,
+        joint_ids,
+        pose_ids,
     )
 
 
 def get_forward_kinematics(robot: Robot, link_names: Optional[List[str]] = None):
-    ForwardKinematics, _, jforward_kinematics, _, _ = ForwardKinematicsFactory(
-        robot, link_names
+    (
+        ForwardKinematics,
+        _,
+        _,
+        backward_helper,
+        link_ids,
+        _,
+        pose_ids,
+    ) = ForwardKinematicsFactory(robot, link_names)
+
+    links = robot.get_links()
+    selected_link_names = [links[id].name for id in pose_ids]
+    SelectedForwardKinematics, _, _, _, _, _, _ = ForwardKinematicsFactory(
+        robot, selected_link_names
     )
     forward_kinematics = ForwardKinematics.apply
-    return forward_kinematics, jforward_kinematics
+
+    def forward_kinematics_body_jacobian(angles: torch.Tensor):
+        selected_poses: torch.Tensor = SelectedForwardKinematics.apply(angles)
+        poses = [None] * robot.num_links
+        for id, pose in zip(pose_ids, selected_poses):
+            poses[id] = pose
+        jposes: torch.Tensor = backward_helper(poses)
+
+        rets = tuple(poses[id] for id in link_ids)
+        jacs: List[torch.Tensor] = []
+
+        for link_id in link_ids:
+            pose = poses[link_id]
+            jac = jposes.new_zeros(angles.shape[0], 6, robot.dof)
+            sel = robot.links[link_id].ancestor_non_fixed_joint_ids
+            jac[:, :, sel] = SE3.adj(SE3.inv(pose)) @ jposes[:, :, sel]
```

Comment:
The body of `forward_kinematics_body_jacobian` is almost the same as that of `forward_kinematics_spatial_jacobian`; only one line changes between the two.  

Can we please stop duplicating code? :'(

Reply:
Implemented.